server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -141,77 +145,112 @@ public void clusterChanged(final ClusterChangedEvent event) {
             return;
         }
 
-        final DiscoveryNodes nodes = state.nodes();
-        if (state.nodes().getMasterNodeId() == null) {
-            logger.debug("not recovering from gateway, no master elected yet");
-        } else if (recoverAfterDataNodes != -1 && nodes.getDataNodes().size() < recoverAfterDataNodes) {
+        if (recoverAfterDataNodes != -1 && nodes.getDataNodes().size() < recoverAfterDataNodes) {
```

Comment:
Maybe move this condition into the per-term check too?

Reply:
I moved the check inside the new `PendingStateRecovery` class before scheduling the recovery. I assume this is what you mean by "per-term", it is *not* about checking it inside `ClusterStateUpdateTask#execute`. The former allows us to keep the same semantics as of today, i.e. no action at all until the required number of data nodes is met.